azure/datalake/store/core.py

```diff
@@ -107,73 +107,101 @@ def open(self, path, mode='rb', blocksize=2**25, delimiter=None):
         return AzureDLFile(self, AzureDLPath(path), mode, blocksize=blocksize,
                            delimiter=delimiter)
 
-    def _ls(self, path):
+    def _ls(self, path, invalidate_cache=True):
         """ List files at given path """
         path = AzureDLPath(path).trim()
         key = path.as_posix()
+        
+        if invalidate_cache:
+            self.invalidate_cache(key)
+
         if key not in self.dirs:
             out = self.azure.call('LISTSTATUS', key)
             self.dirs[key] = out['FileStatuses']['FileStatus']
             for f in self.dirs[key]:
                 f['name'] = (path / f['pathSuffix']).as_posix()
         return self.dirs[key]
 
-    def ls(self, path="", detail=False):
+    def ls(self, path="", detail=False, invalidate_cache=True):
         """ List single directory with or without details """
         path = AzureDLPath(path)
-        files = self._ls(path)
+        files = self._ls(path, invalidate_cache)
         if not files:
-            inf = self.info(path)
+            # in this case we just invalidated the cache (if it was true), so no need to do it again
+            inf = self.info(path, invalidate_cache=False)
             if inf['type'] == 'DIRECTORY':
-                return []
+                return inf if detail else []
+
             raise FileNotFoundError(path)
         if detail:
             return files
         else:
             return [f['name'] for f in files]
 
-    def info(self, path):
+    def info(self, path, invalidate_cache=True, expected_error_code=None):
         """ File information
         """
         path = AzureDLPath(path).trim()
-        root = path.parent
         path_as_posix = path.as_posix()
-        for f in self._ls(root):
+        root = path.parent
+        root_as_posix = root.as_posix()
+        
+        # in the case of getting info about the root itself or if the cache won't be hit
+        # simply return the result of a GETFILESTATUS from the service
+        if invalidate_cache or path_as_posix == '/' or path_as_posix == '.':
+            to_return  = self.azure.call('GETFILESTATUS', path_as_posix, expected_error_code=expected_error_code)['FileStatus']
+            to_return['name'] = path_as_posix
+            
+            # add the key/value pair back to the cache so long as it isn't the root
+            if path_as_posix != '/' and path_as_posix != '.':
+                if root_as_posix not in self.dirs:
+                    self.dirs[root_as_posix] = [to_return]
+                else:
+                    found = False
+                    for f in self.dirs[root_as_posix]:
+                        if f['name'] == path_as_posix:
+                            found = True
+                            f = to_return
+                    if not found:
+                        self.dirs[root_as_posix].append(to_return)
```

Comment:
good catch, I missed the break previously. We are looking for the first f, since the name property is unique within a directory (we are basically going through a list of files/folders in a specific directory and looking for the one that matches).